src/step-function.ts

```diff
@@ -466,6 +465,152 @@ export namespace $SFN {
+   *   const results = $SFN.retry(
+   *     async () => { await task1(id) }
+   *   )
+   * })
+   * ```
+   *
+   * By default, all errors (`States.ALL`) will be caught and retried with the default backoff strategy:
+   * MaxAttempts: 3
+   * IntervalSeconds: 1
+   * BackoffRate: 2.0
+   *
+   * Optionally provide specific error codes and backoff strategies.
+   *
+   * ```ts
+   * new ExpressStepFunction(this, "F", (id: string) => {
+   *   const results = $SFN.retry(
+   *     [{ ErrorEquals: ["Lambda.ToManyRequestsException"], BackoffRate: 2.5, IntervalSeconds: 2, MaxAttempts: 4 }],
+   *     async () => { await task1(id) }
+   *   )
+   * })
+   * ```
+   *
+   * Caveat: `$SFN.retry` does not maintain the state of mutated variables within the callback.
+   *
+   * ```ts
+   * let a = 1;
+   * await $SFN.retry(() => { a = 2; return a * 4; })
+   * return a;
+   * ```
+   *
+   * The return value will be `1`.
+   *
+   */
+  export const retry = makeStepFunctionIntegration<"$SFN.retry", RetryFunc>(
+    "$SFN.retry",
+    {
+      asl(call, context) {
+        const [retries, body] =
+          call.args.length === 1 ? [undefined, call.args[0]] : call.args;
+
+        if (!body || !isFunctionLike(body.expr)) {
+          throw new SynthError(
+            ErrorCodes.Invalid_Input,
+            "Expected $SFN.retry callback argument to be a function."
+          );
+        }
+
+        const retryValues = processRetries(retries?.expr);
+
+        const functionStates = context.evalStmt(body.expr.body, {
+          End: true,
+          ResultPath: "$",
+        });
+
+        /**
+         * If there are no states for the function, just return an empty state, nothing to do here.
+         */
+        if (!functionStates) {
+          return {
+            Type: "Pass",
+            Next: ASLGraph.DeferNext,
+            output: { value: context.context.null, containsJsonPath: false },
+          };
+        }
+
+        const temp = context.newHeapVariable();
+
+        return {
+          Type: "Parallel",
+          Branches: [context.aslGraphToStates(functionStates)],
+          Next: ASLGraph.DeferNext,
+          Retry: retryValues,
+          ResultPath: temp,
+          Parameters: context.cloneLexicalScopeParameters(body.expr),
+          output: {
+            jsonPath: `${temp}[0]`,
+          },
+        };
+
+        function processRetries(expr?: Expr): Retry[] {
+          if (!expr) {
+            return [
+              {
+                ErrorEquals: ["States.ALL"],
+              },
+            ];
+          }
+
+          const retryValue = evalToConstant(expr);
+
+          if (
+            !retryValue ||
+            !Array.isArray(retryValue.constant) ||
+            retryValue.constant.length === 0
+          ) {
+            throw new SynthError(
+              ErrorCodes.Invalid_Input,
+              "Expected $SFN.retry retry argument to be an array literal which contains only constant values."
```

Comment:
The error message doesn't explain that the array can't be empty.

I think this case deserves its own ErrorCode with explanation.

Reply:
I have been trying to reduce error codes related to specific integrations, but I guess this is a common enough case.